test/compiler.test.js

```diff
@@ -1014,6 +1014,201 @@ describe('compiler', function() {
+        var instructions = boot.compile(options);
+
+        expect(instructions.mixins).to.have.property('FOO');
+        expect(instructions.mixins).to.have.property('TIME-STAMPS');
+      });
+
+      it('skips mixin name normalization - none', function() {
+        appdir.writeFileSync('mixins/foo.js', '');
+        appdir.writeFileSync('mixins/time-stamps.js', '');
+
+        var options = { appRootDir: appdir.PATH, normalization: 'none' };
+        var instructions = boot.compile(options);
+
+        expect(instructions.mixins).to.have.property('foo');
+        expect(instructions.mixins).to.have.property('time-stamps');
+      });
+
+      it('skips mixin name normalization - false', function() {
+        appdir.writeFileSync('mixins/foo.js', '');
+        appdir.writeFileSync('mixins/time-stamps.js', '');
+
+        var options = { appRootDir: appdir.PATH, normalization: false };
+        var instructions = boot.compile(options);
+
+        expect(instructions.mixins).to.have.property('foo');
+        expect(instructions.mixins).to.have.property('time-stamps');
+      });
+
+      it('defaults normalization format to - classify', function() {
+        appdir.writeFileSync('mixins/foo.js', '');
+        appdir.writeFileSync('mixins/time-stamps.js', '');
+
+        var options = { appRootDir: appdir.PATH };
+        var instructions = boot.compile(options);
+
+        expect(instructions.mixins).to.have.property('Foo');
+        expect(instructions.mixins).to.have.property('TimeStamps');
+      });
+
+      it('throws error for invalid normalization format', function() {
+        appdir.writeFileSync('mixins/foo.js', '');
+
+        var options = { appRootDir: appdir.PATH,
+          normalization: 'invalidFormat'
+        };
+
+        expect(function() { boot.compile(options); })
+        .to.throw(/Invalid normalization format - "invalidFormat"/);
+      });
+
+      it('overrides default mixin name, by `name` in JSON', function() {
+        appdir.writeFileSync('mixins/foo.js', '');
+        appdir.writeConfigFileSync('mixins/foo.json', {name: 'fooBar'});
+
+        var instructions = boot.compile(appdir.PATH);
+
+        expect(instructions.mixins).to.have.property('fooBar');
+      });
+
+      it('extends definition from JSON with same file name', function() {
+        var appJS = appdir.writeFileSync('mixins/foo-bar.js', '');
+
+        appdir.writeConfigFileSync('mixins/foo-bar.json', {
+          description: 'JSON file name same as JS file name' });
+        appdir.writeConfigFileSync('mixins/FooBar.json', {
+          description: 'JSON file name same as normalized name of mixin' });
+
+        var options = { appRootDir: appdir.PATH, normalization: 'classify' };
+        var instructions = boot.compile(options);
+
+        expect(instructions.mixins).to.eql({
+          FooBar: {
+            name: 'FooBar',
+            description: 'JSON file name same as JS file name',
+            filepath: appJS
+          }
+        });
```

Comment:
I am concerned about mixing metadata from the JSON file with boot data like `filepath`. I would rather use the same approach that we use for models, see [this code](https://github.com/PradnyaBaviskar/loopback-boot/blob/1b9f13706229a80877012428fceeda6dbea85356/lib/compiler.js#L203-L208). Also `filepath` should be called `sourceFile`.